components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt

```diff
@@ -76,74 +76,81 @@ class AppLinksUseCases(
+            val fallbackUrl = if (redirectData.fallbackIntent?.data?.isHttpOrHttps == true) {
+                redirectData.fallbackIntent.dataString
+            } else {
+                null
+            }
 
-            val appInfo = resolveInfos?.firstOrNull()
+            return AppLinkRedirect(appIntent, fallbackUrl, redirectData.resolveInfo)
+        }
 
-            return AppLinkRedirect(appIntent, webUrl, webUrl != url, appInfo)
+        private fun isDefaultBrowser(intent: Intent?): Boolean {
+            return when (intent) {
+                null -> false
+                else -> findDefaultActivity(intent)
+                        ?.activityInfo?.packageName == context.packageName
+            }
         }
 
-        private fun getNonBrowserActivities(intent: Intent): List<ResolveInfo> {
+        private fun getNonBrowserActivities(intent: Intent?): List<ResolveInfo> {
+            if (intent == null) {
+                return listOf()
+            }
+
             return findActivities(intent)
                 .map { it.activityInfo.packageName to it }
                 .filter { !browserPackageNames.contains(it.first) }
                 .map { it.second }
         }
 
-        private fun createBrowsableIntents(url: String): List<Intent> {
+        private fun createBrowsableIntents(url: String): RedirectData {
             val intent = Intent.parseUri(url, 0)
 
             if (intent.action == Intent.ACTION_VIEW) {
                 intent.addCategory(Intent.CATEGORY_BROWSABLE)
+                intent.component = null
+                intent.selector = null
             }
 
-            return when (intent.data?.isHttpOrHttps) {
-                null -> emptyList()
-                true -> listOf(intent)
-                false -> {
-                    // Non http[s] schemes:
-
-                    val fallback = intent.getStringExtra(EXTRA_BROWSER_FALLBACK_URL)?.let {
-                        Intent.parseUri(it, 0)
-                    }
-
-                    val marketplaceIntent = intent.`package`?.let {
-                        if (includeInstallAppFallback) {
-                            Intent.parseUri(MARKET_INTENT_URI_PACKAGE_PREFIX + it, 0)
-                        } else {
-                            null
-                        }
-                    }
-
-                    return listOfNotNull(intent, fallback, marketplaceIntent)
+            val fallbackIntent = intent.getStringExtra(EXTRA_BROWSER_FALLBACK_URL)?.let {
+                Intent.parseUri(it, 0)
+            }
+
+            val marketplaceIntent = intent.`package`?.let {
+                if (includeInstallAppFallback) {
+                    Intent.parseUri(MARKET_INTENT_URI_PACKAGE_PREFIX + it, 0)
+                } else {
+                    null
                 }
             }
+
+            val appIntent = when (intent.data?.isHttpOrHttps) {
+                null -> null
+                else -> intent
+            }
```

Comment:
This looks weird. If we've decided it's `http` or `https`, then it's the `appIntent`, otherwise, it's `null`?

i.e.

```kotlin
val appIntent = intent.data?.isHttpOrHttps?.let {
    intent
}
```

If that's the intended logic, then could you comment why it looks weird?

Reply:
That is the intent (pun not intended).  Your suggestion is much cleaner, I'll update. Thanks,